File: advent/day23.cc
#include<cstdlib>
#include<iostream>


int main() {
    const size_t NUM_VALS = 1000000;
    const uint64_t NUM_ROUNDS = 10000000;

    uint64_t *next = (uint64_t*) malloc(NUM_VALS * sizeof(uint64_t));
    next[2 - 1] = 1 - 1;
    next[1 - 1] = 5 - 1;
    next[5 - 1] = 6 - 1;
    next[6 - 1] = 9 - 1;
    next[9 - 1] = 4 - 1;
    next[4 - 1] = 7 - 1;
    next[7 - 1] = 8 - 1;
    next[8 - 1] = 3 - 1;
    next[3 - 1] = 9;

    for (size_t i = 9; i < NUM_VALS; ++i) {
        next[i] = i + 1;
    }
    next[NUM_VALS - 1] = 2 - 1;

    uint64_t cur_cup = 2 - 1;
    for (u_int64_t round_num = 0; round_num < NUM_ROUNDS; ++round_num) {

        const uint64_t first = next[cur_cup],
                       second = next[first],
                       third = next[second];

        uint64_t next_cup = (cur_cup > 0) ? ((cur_cup - 1) % NUM_VALS) : (NUM_VALS - 1);
        while (next_cup == first || next_cup == second || next_cup == third) {
            next_cup = (next_cup > 0) ? ((next_cup - 1) % NUM_VALS) : (NUM_VALS - 1);
        }

        next[cur_cup] = next[third];
        uint64_t tmp = next[next_cup];
        next[next_cup] = first;
        next[third] = tmp;

        cur_cup = next[cur_cup];
    }

    std::cout << ((next[0] + 1) * (next[next[0]] + 1)) << std::endl;
}

File: advent/day23.py
""" AOC Day 23 """
from pathlib import Path
from typing import List, Union


def first(filename: Union[str, Path], num_rounds: int = 100) -> int:
    """
    Part 1
    """
    with open(filename, "rt") as infile:
        cups = [int(x) for x in infile.read().strip()]

    for num_round in range(num_rounds):
        cur_idx = 0
        cur_idx = num_round % len(cups)
        cur_cup = cups[cur_idx]

        removed_cups = [cups[(cur_idx + i) % len(cups)] for i in range(1, 4)]

        next_cup = ((cur_cup - 1) + len(cups)) % len(cups)
        if next_cup == 0:
            next_cup = len(cups)

        while next_cup in removed_cups:
            next_cup -= 1
            if next_cup == 0:
                next_cup = len(cups)

        next_cups: List[int] = []
        while len(next_cups) < len(cups):
            this_cup = cups[cur_idx]
            if this_cup in removed_cups:
                pass
            elif this_cup == next_cup:
                next_cups.append(next_cup)
                next_cups.extend(removed_cups)
            else:
                next_cups.append(this_cup)

            cur_idx = (cur_idx + 1) % len(cups)

        cups = (
            next_cups[-(num_round % len(cups)) :]
            + next_cups[: -(num_round % len(cups))]
        )

    idx = cups.index(1)
    output = []
    cur_idx = (idx + 1) % len(cups)
    while cur_idx != idx:
        output.append(cups[cur_idx])
        cur_idx += 1
        cur_idx %= len(cups)

    return int("".join(map(str, output)))


def second(filename: Union[str, Path], num_rounds: int = 10_000_000) -> int:
    """
    Part 2
    """
    vals: List[int] = []
    with open(filename, "rt") as infile:
        for cup in infile.read().strip():
            vals.append(int(cup) - 1)

    vals.extend(range(max(vals) + 1, 1_000_000))

    num_vals = 1_000_000
    val_to_next = [0] * num_vals
    for val, next_val in zip(vals[:-1], vals[1:]):
        val_to_next[val] = next_val
    val_to_next[vals[-1]] = vals[0]

    cur_val = vals[0]

    for _ in range(num_rounds):
        removed_vals = [
            val_to_next[cur_val],
            val_to_next[val_to_next[cur_val]],
            val_to_next[val_to_next[val_to_next[cur_val]]],
        ]

        next_cup = (cur_val - 1) % num_vals
        while next_cup in removed_vals:
            next_cup = (next_cup - 1) % num_vals

        val_to_next[cur_val] = val_to_next[removed_vals[-1]]
        new_next = val_to_next[next_cup]
        val_to_next[next_cup] = removed_vals[0]
        val_to_next[removed_vals[-1]] = new_next

        cur_val = val_to_next[cur_val]

    return (val_to_next[0] + 1) * (val_to_next[val_to_next[0]] + 1)
